File: src/main/java/io/github/numichi/reactive/logger/DefaultValues.java
package io.github.numichi.reactive.logger;

import io.github.numichi.reactive.logger.annotations.JacocoSkipGeneratedReport;
import io.github.numichi.reactive.logger.exception.AlreadyConfigurationException;
import reactor.core.scheduler.Scheduler;
import reactor.core.scheduler.Schedulers;

@JacocoSkipGeneratedReport
public class DefaultValues {
    private static final String DEFAULT_REACTOR_CONTEXT_MDC_KEY = "EFAULT_REACTOR_CONTEXT_MDC_KEY";
    private static final Scheduler DEFAULT_SCHEDULER = Schedulers.boundedElastic();
    private static DefaultValues instance = null;
    private final String defaultReactorContextMdcKey;
    private final Scheduler defaultScheduler;
    
    private DefaultValues() {
        this(DEFAULT_REACTOR_CONTEXT_MDC_KEY, DEFAULT_SCHEDULER);
    }
    
    private DefaultValues(Scheduler defaultScheduler) {
        this(DEFAULT_REACTOR_CONTEXT_MDC_KEY, defaultScheduler);
    }
    
    private DefaultValues(String defaultReactorContextMdcKey) {
        this(defaultReactorContextMdcKey, DEFAULT_SCHEDULER);
    }
    
    private DefaultValues(String defaultReactorContextMdcKey, Scheduler defaultScheduler) {
        this.defaultReactorContextMdcKey = defaultReactorContextMdcKey;
        this.defaultScheduler = defaultScheduler;
    }
    
    public static DefaultValues getInstance() {
        if (instance == null) {
            instance = configuration();
        }
        
        return instance;
    }
    
    public static DefaultValues configuration() throws AlreadyConfigurationException {
        if (instance == null) {
            instance = new DefaultValues();
            return instance;
        } else {
            throw new AlreadyConfigurationException();
        }
    }
    
    public static DefaultValues configuration(Scheduler defaultScheduler) throws AlreadyConfigurationException {
        if (instance == null) {
            instance = new DefaultValues(defaultScheduler);
            return instance;
        } else {
            throw new AlreadyConfigurationException();
        }
    }
    
    public static DefaultValues configuration(String defaultReactorContextMdcKey) throws AlreadyConfigurationException {
        if (instance == null) {
            instance = new DefaultValues(defaultReactorContextMdcKey);
            return instance;
        } else {
            throw new AlreadyConfigurationException();
        }
    }
    
    public static DefaultValues configuration(String defaultReactorContextMdcKey, Scheduler defaultScheduler) throws AlreadyConfigurationException {
        if (instance == null) {
            instance = new DefaultValues(defaultReactorContextMdcKey, defaultScheduler);
            return instance;
        } else {
            throw new AlreadyConfigurationException();
        }
    }
    
    public String getDefaultReactorContextMdcKey() {
        return defaultReactorContextMdcKey;
    }
    
    public Scheduler getDefaultScheduler() {
        return defaultScheduler;
    }
}

File: src/main/java/io/github/numichi/reactive/logger/DefaultValues.kt
package io.github.numichi.reactive.logger

import io.github.numichi.reactive.logger.exception.AlreadyConfigurationException
import io.github.numichi.reactive.logger.models.MDCHook
import reactor.core.scheduler.Scheduler
import reactor.core.scheduler.Schedulers

class DefaultValues private constructor(
    var defaultReactorContextMdcKey: String = DEFAULT_REACTOR_CONTEXT_MDC_KEY,
    var defaultScheduler: Scheduler = Schedulers.boundedElastic(),
    var customHook: MutableList<MDCHook<*>> = mutableListOf()
) {

    @Throws(AlreadyConfigurationException::class)
    fun addHook(hook: MDCHook<*>) {
        customHook.add(hook)
    }

    fun removeHook(filter: (contextKey: Any) -> Boolean) {
        customHook = customHook.filter { filter(it.contextKey) }.toMutableList()
    }

    fun reset() {
        defaultReactorContextMdcKey = DEFAULT_REACTOR_CONTEXT_MDC_KEY
        defaultScheduler = Schedulers.boundedElastic()
        customHook = mutableListOf()
    }

    companion object {
        private var singleton: DefaultValues? = null

        @JvmStatic
        fun getInstance(): DefaultValues {
            if (singleton == null) {
                singleton = DefaultValues()
            }

            return singleton!!
        }
    }
}